src/webapi/rendering_context.rs

```diff
@@ -18,6 +20,160 @@ pub trait RenderingContext {
+#[reference(instance_of = "CanvasPattern")]
+pub struct CanvasPattern(Reference);
+
+/// The ImageData interface represents the underlying pixel data of an area of a <canvas> element. 
+/// It is created using the ImageData() constructor or creator methods on the CanvasRenderingContext2D
+///  object associated with a canvas: createImageData() and getImageData(). It can also be used to set 
+/// a part of the canvas by using putImageData().
+/// 
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/ImageData)
+// https://html.spec.whatwg.org/#imagedata
+#[derive(Clone, Debug, ReferenceType)]
+#[reference(instance_of = "ImageData")]
+pub struct ImageData(Reference);
+
+/// The TextMetrics interface represents the dimension of a text in the canvas, as created by the CanvasRenderingContext2D.measureText() method.
+/// 
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/TextMetrics)
+// https://html.spec.whatwg.org/#textmetrics
+#[derive(Clone, Debug, ReferenceType)]
+#[reference(instance_of = "TextMetrics")]
+pub struct TextMetrics(Reference);
+
+/// The type of compositing operation to apply when drawing new shapes, where type is a string 
+/// identifying which of the twelve compositing operations to use.
+/// 
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/globalCompositeOperation)
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum CompositeOperation {
+    SourceOver,
+    SourceIn,
+    SourceOut,
+    SourceAtop,
+    DestinationOver,
+    DestinationIn,
+    DestinationAtop,
+    Lighter,
+    Copy,
+    Xor,
+    Multiply,
+    Screen,
+    Overlay,
+    Darken,
+    Lighten,
+    ColorDodge,
+    ColorBurn,
+    HardLight,
+    SoftLight,
+    Difference,
+    Exclusion,
+    Hue,
+    Saturation,
+    Color,
+    Luminosity
+}
+
+/// The algorithm by which to determine if a point is inside a path or outside a path.
+/// 
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/fill)
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum FillRule {
+    NonZero,
+    EvenOdd
+}
+
+/// How the end points of every line are drawn.
+/// 
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/lineCap)
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum LineCap {
+    Butt,
+    Round,
+    Square
+}
+
+/// determines how two connecting segments (of lines, arcs or curves) with non-zero lengths in a shape are 
+/// joined together (degenerate segments with zero lengths, whose specified endpoints and control points are 
+/// exactly at the same position, are skipped).
+/// 
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/lineJoin)
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum LineJoin {
+    Bevel,
+    Round,
+    Miter
+}
+
+/// A DOMString indicating how to repeat the image.
```

Comment:
Minor nitpick: `DOMString` part is irrelevant, since in Rust this is an enum. (: